False Positive on contextmanager-generator-missing-cleanup / W0135 When Using With Expression in the ContextManager

### Bug description
The `contextmanager-generator-missing-cleanup` warning triggers a false positive (by my understanding of the warning) when a `yield` expression is fully cleaned up by a surrounding `with` statement (or multiple with statements). 

Below is a simplified example of the issue:

```python
import contextlib


@contextlib.contextmanager
def cm():
    with open("/tmp/test", "wb+") as contextvar:
        yield contextvar.fileno()


def genfunc_with_cm():  # [contextmanager-generator-missing-cleanup]
    with cm() as context:
        yield context * 2
```

The warning can be silenced with the guidance from https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/contextmanager-generator-missing-cleanup.html . But by my understanding this is an unnecessary warning in this case. The `with` statement in `cm()` does the 'right thing' when `yield` throws `GeneratorExit` and closes the file.


### Configuration

_No response_

### Command used

```shell
pylint ./test.py
```


### Pylint output

```python
************* Module test
W  test.py:10: The context used in function 'genfunc_with_cm' will not be exited. (contextmanager-generator-missing-cleanup | W0135)
```


### Expected behavior

I would expect the `contextmanager-generator-missing-cleanup` to only warn when there is extra cleanup required in a `contextmanager` outside of  `with` statements. 

I think you see the intention of these cases being supported in the `good_cm_yield_none()` example in https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/contextmanager-generator-missing-cleanup.html

If there's nothing to clean up: the warning should not be generated.

### Pylint version

```shell
pylint 3.2.0
astroid 3.2.0
Python 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
```


### OS / Environment

Ubuntu 22.04 LTS

### Additional dependencies

_No response_

Similar case: If I have code that is decorated with `@contextmanager`, creates `with TemporaryDirectory()` and does some prep work, then exits the function by yielding a file in the temporary directory -- there is no expected code to run after yield but I do want the parent tempdir cleanup.

This check seems to indicate that I should explicitly wrap it in try/finally with cleanup code consisting of "pass", which feels wrong because that's just a pinky promise that cleanup is done.

```
mesonbuild/compilers/compilers.py:858:4: W0135: The context used in function 'cached_compile' will not be exited. (contextmanager-generator-missing-cleanup)
```

--- a/doc/data/messages/c/contextmanager-generator-missing-cleanup/bad.py
+++ b/doc/data/messages/c/contextmanager-generator-missing-cleanup/bad.py
@@ -9,6 +9,6 @@ def cm():
     print("cm exit")
 
 
-def genfunc_with_cm():  # [contextmanager-generator-missing-cleanup]
-    with cm() as context:
+def genfunc_with_cm():
+    with cm() as context:  # [contextmanager-generator-missing-cleanup]
         yield context * 2

--- a/doc/data/messages/c/contextmanager-generator-missing-cleanup/details.rst
+++ b/doc/data/messages/c/contextmanager-generator-missing-cleanup/details.rst
@@ -8,3 +8,24 @@ because the ways to use a contextmanager are many.
 A contextmanager can be used as a decorator (which immediately has ``__enter__``/``__exit__`` applied)
 and the use of ``as ...`` or discard of the return value also implies whether the context needs cleanup or not.
 So for this message, warning the invoker of the contextmanager is important.
+
+The check can create false positives if ``yield`` is used inside an ``if-else`` block without custom cleanup. Use ``pylint: disable`` for these.
+
+.. code-block:: python
+
+    from contextlib import contextmanager
+
+    @contextmanager
+    def good_cm_no_cleanup():
+        contextvar = "acquired context"
+        print("cm enter")
+        if some_condition:
+            yield contextvar
+        else:
+            yield contextvar
+
+
+    def good_cm_no_cleanup_genfunc():
+        # pylint: disable-next=contextmanager-generator-missing-cleanup
+        with good_cm_no_cleanup() as context:
+            yield context * 2

--- a/doc/data/messages/c/contextmanager-generator-missing-cleanup/good.py
+++ b/doc/data/messages/c/contextmanager-generator-missing-cleanup/good.py
@@ -47,3 +47,15 @@ def good_cm_finally():
 def good_cm_finally_genfunc():
     with good_cm_finally() as context:
         yield context * 2
+
+
+@contextlib.contextmanager
+def good_cm_no_cleanup():
+    contextvar = "acquired context"
+    print("cm enter")
+    yield contextvar
+
+
+def good_cm_no_cleanup_genfunc():
+    with good_cm_no_cleanup() as context:
+        yield context * 2

--- a/doc/whatsnew/fragments/9625.false_positive
+++ b/doc/whatsnew/fragments/9625.false_positive
@@ -1,0 +1,4 @@
+Exclude context manager without cleanup from
+``contextmanager-generator-missing-cleanup`` checks.
+
+Closes #9625

--- a/pylint/checkers/base/function_checker.py
+++ b/pylint/checkers/base/function_checker.py
@@ -72,7 +72,7 @@ class FunctionChecker(_BasicChecker):
                 if self._node_fails_contextmanager_cleanup(inferred_node, yield_nodes):
                     self.add_message(
                         "contextmanager-generator-missing-cleanup",
-                        node=node,
+                        node=with_node,
                         args=(node.name,),
                     )
 

--- a/pylint/checkers/base/function_checker.py
+++ b/pylint/checkers/base/function_checker.py
@@ -85,6 +85,7 @@ class FunctionChecker(_BasicChecker):
         Current checks for a contextmanager:
             - only if the context manager yields a non-constant value
             - only if the context manager lacks a finally, or does not catch GeneratorExit
+            - only if some statement follows the yield, some manually cleanup happens
 
         :param node: Node to check
         :type node: nodes.FunctionDef
@@ -114,6 +115,19 @@ class FunctionChecker(_BasicChecker):
             for yield_node in yield_nodes
         ):
             return False
+
+        # Check if yield expression is last statement
+        yield_nodes = list(node.nodes_of_class(nodes.Yield))
+        if len(yield_nodes) == 1:
+            n = yield_nodes[0].parent
+            while n is not node:
+                if n.next_sibling() is not None:
+                    break
+                n = n.parent
+            else:
+                # No next statement found
+                return False
+
         # if function body has multiple Try, filter down to the ones that have a yield node
         try_with_yield_nodes = [
             try_node

--- a/tests/functional/c/contextmanager_generator_missing_cleanup.py
+++ b/tests/functional/c/contextmanager_generator_missing_cleanup.py
@@ -175,3 +175,15 @@ def genfunc_with_cm_bare_handler():
 def genfunc_with_cm_base_exception_handler():
     with cm_base_exception_handler() as context:
         yield context * 2
+
+
+@contextlib.contextmanager
+def good_cm_no_cleanup():
+    contextvar = "acquired context"
+    print("cm enter")
+    yield contextvar
+
+
+def good_cm_no_cleanup_genfunc():
+    with good_cm_no_cleanup() as context:
+        yield context * 2

--- a/tests/functional/c/contextmanager_generator_missing_cleanup.py
+++ b/tests/functional/c/contextmanager_generator_missing_cleanup.py
@@ -14,8 +14,8 @@ def cm():
     print("cm exit")
 
 
-def genfunc_with_cm():  # [contextmanager-generator-missing-cleanup]
-    with cm() as context:
+def genfunc_with_cm():
+    with cm() as context:  # [contextmanager-generator-missing-cleanup]
         yield context * 2
 
 
@@ -27,13 +27,13 @@ def name_cm():
     print("cm exit")
 
 
-def genfunc_with_name_cm():  # [contextmanager-generator-missing-cleanup]
-    with name_cm() as context:
+def genfunc_with_name_cm():
+    with name_cm() as context:  # [contextmanager-generator-missing-cleanup]
         yield context * 2
 
 
-def genfunc_with_cm_after():  # [contextmanager-generator-missing-cleanup]
-    with after_cm() as context:
+def genfunc_with_cm_after():
+    with after_cm() as context:  # [contextmanager-generator-missing-cleanup]
         yield context * 2
 
 
@@ -56,8 +56,8 @@ def cm_with_improper_handling():
     print("cm exit")
 
 
-def genfunc_with_cm_improper():  # [contextmanager-generator-missing-cleanup]
-    with cm_with_improper_handling() as context:
+def genfunc_with_cm_improper():
+    with cm_with_improper_handling() as context:  # [contextmanager-generator-missing-cleanup]
         yield context * 2
 
 

--- a/tests/functional/c/contextmanager_generator_missing_cleanup.txt
+++ b/tests/functional/c/contextmanager_generator_missing_cleanup.txt
@@ -1,4 +1,4 @@
-contextmanager-generator-missing-cleanup:17:0:17:19:genfunc_with_cm:The context used in function 'genfunc_with_cm' will not be exited.:UNDEFINED
-contextmanager-generator-missing-cleanup:30:0:30:24:genfunc_with_name_cm:The context used in function 'genfunc_with_name_cm' will not be exited.:UNDEFINED
-contextmanager-generator-missing-cleanup:35:0:35:25:genfunc_with_cm_after:The context used in function 'genfunc_with_cm_after' will not be exited.:UNDEFINED
-contextmanager-generator-missing-cleanup:59:0:59:28:genfunc_with_cm_improper:The context used in function 'genfunc_with_cm_improper' will not be exited.:UNDEFINED
+contextmanager-generator-missing-cleanup:18:4:19:25:genfunc_with_cm:The context used in function 'genfunc_with_cm' will not be exited.:UNDEFINED
+contextmanager-generator-missing-cleanup:31:4:32:25:genfunc_with_name_cm:The context used in function 'genfunc_with_name_cm' will not be exited.:UNDEFINED
+contextmanager-generator-missing-cleanup:36:4:37:25:genfunc_with_cm_after:The context used in function 'genfunc_with_cm_after' will not be exited.:UNDEFINED
+contextmanager-generator-missing-cleanup:60:4:61:25:genfunc_with_cm_improper:The context used in function 'genfunc_with_cm_improper' will not be exited.:UNDEFINED

